File: utils/src/main/java/ru/virgil/utils/image/ImageMockService.java
package ru.virgil.utils.image;

import lombok.RequiredArgsConstructor;
import org.springframework.mock.web.MockMultipartFile;
import ru.virgil.utils.FakerUtils;

import javax.annotation.PreDestroy;
import java.io.BufferedInputStream;
import java.io.IOException;
import java.net.URL;

@RequiredArgsConstructor
public abstract class ImageMockService<Owner extends IBaseEntity, Image extends IPrivateImage<Owner>> {

    public static final String IMAGE_NAME = "image";
    protected final ImageService<Owner, Image> imageService;
    protected final FakerUtils fakerUtils;
    protected final ImageProperties imageProperties;

    public Image mockImage(Owner owner) {
        try {
            return imageService.savePrivate(mockAsMultipart().getBytes(), IMAGE_NAME, owner);
        } catch (IOException e) {
            throw new ImageException(e);
        }
    }

    public MockMultipartFile mockAsMultipart() {
        try {
            BufferedInputStream in = new BufferedInputStream(new URL(fakerUtils.avatar().image()).openStream());
            return new MockMultipartFile(IMAGE_NAME, in);
        } catch (IOException e) {
            throw new ImageException(e);
        }
    }

    @PreDestroy
    public void preDestroy() throws IOException {
        if (imageProperties.cleanOnDestroy()) {
            imageService.cleanFolders();
        }
    }
}

File: utils/src/main/java/ru/virgil/utils/image/ImageMockService.kt
package ru.virgil.utils.image

import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.mock.web.MockMultipartFile
import ru.virgil.utils.Faker
import ru.virgil.utils.base.entity.Identified
import java.io.BufferedInputStream
import java.io.IOException
import java.net.URL
import javax.annotation.PreDestroy

@EnableConfigurationProperties(ImageProperties::class)
abstract class ImageMockService<Owner : Identified, Image : PrivateImage<Owner>>(
    @Suppress("MemberVisibilityCanBePrivate")
    protected val imageService: ImageService<Owner, Image>,
    @Suppress("MemberVisibilityCanBePrivate")
    protected val fakerUtils: Faker,
    @Suppress("MemberVisibilityCanBePrivate")
    protected val imageProperties: ImageProperties,
) {

    @Throws(ImageException::class)
    fun mockImage(owner: Owner): Image = try {
        imageService.savePrivate(mockAsMultipart().bytes, IMAGE_NAME, owner)
    } catch (e: IOException) {
        throw ImageException(e)
    }

    @Throws(ImageException::class)
    fun mockAsMultipart(): MockMultipartFile = try {
        val bufferedInputStream = BufferedInputStream(URL(fakerUtils.avatar().image()).openStream())
        MockMultipartFile(IMAGE_NAME, bufferedInputStream)
    } catch (e: IOException) {
        throw ImageException(e)
    }

    @PreDestroy
    @Throws(IOException::class)
    fun preDestroy() {
        if (imageProperties.cleanOnDestroy) {
            imageService.cleanFolders()
        }
    }

    companion object {

        const val IMAGE_NAME = "image"
    }
}
